go/private/tools/path.bzl

```diff
@@ -144,10 +169,13 @@ def _go_path_impl(ctx):
         ),
     ]
 
-go_path = rule(
+go_path = go_rule(
     _go_path_impl,
     attrs = {
-        "deps": attr.label_list(providers = [GoArchive]),
+        "deps": attr.label_list(
+            providers = [GoArchive],
+            aspects = [go_archive_aspect],
```

Comment:
I think `go_path` should accept archives in multiple modes. Ideally, you'd be able to point to multiple binaries and libraries with their own modes in `deps`, and `go_path` would include them all in the appropriate directories. That may not actually work at this point (the same archive in multiple modes will probably cause conflicts right now), and that's fine for the PR, but I don't want to commit to a single mode in this API.

In order the make this work, I think you'll need to copy `go.mode` into `GoArchive`. It shouldn't need to be in `GoArchiveData`, since anything an archive depends on will be in the same mode. You won't need an aspect or any mode attributes in this rule.

Reply:
How do you feel this working file layout wise for two libraries in different mode?

Reply:
If you have the same library in multiple modes, then there should be multiple elements in `archives` for that library.

For the sources and data, I don't think we'll actually get conflicts because `_add_manifest_entry` checks whether file paths are different before reporting an error.

For the archives, .a files built in different modes will go in different `pkg` directories. For example, you might have `pkg/linux_amd64/foo.a` and `pkg/linux_amd64_race/foo.a`.

Reply:
That won't work since pkg has a pretty rigid layout and expects GOOS_GOARCH, and likely most tools that depend on it such as gocode.

Also, when setting up code completion we run the target once and once in a while refresh it to update it. Then we point our editor's go_path to the go_path inside bazel-out.

Reply:
Unfortunately most ides (vscode in our case) will boot gocode with a predefined goos and goarch but I don't think there more to it.

Also, I think it would be simpler to have multiple targets, should one need it?

Reply:
> Unfortunately most ides (vscode in our case) will boot gocode with a predefined goos and goarch but I don't think there more to it.

What problems does that cause? This would cause issues if you are building a `go_path` for one target but gocode expects another. If `go_path` depends on a `go_binary` with explicit `goos` and `goarch` attributes, that could cause issues though.

> Also, I think it would be simpler to have multiple targets, should one need it?

You mean multiple directories in `pkg`? I agree, we should support this as much as possible. However, some of the mode attributes (`static`, `pure`, `debug`, `stripped`) don't correspond to any conventional install suffix. So when we're generating the install suffix under pkg, we can either leave them out (lossy), or we can include them but some tools won't expect those parts of the suffix.

Reply:
True...
Well, since this rule is precisely for those tools, the lossy option looks fine to me, as they don't expect a suffix.
Is there a case for bazel tools to use the generated gopath save for rules wrapping some go tools?
In our case we are wrapping `gomobile bind` with success.

Reply:
Sorry I meant to answer this earlier. `go_path` has a couple different purposes.

One is wrapping tools that expect a GOPATH. We have an ongoing project to build a workspace abstraction layer (not Bazel-specific) that will address these kinds of problems for tools in general. So at some point we won't need `go_path` for that.

The other important use case is building source archives. For example, if you have a Bazel project with a non-standard layout or a lot of generated code, you could package everything up with a `go_path` and drop it in a vendor directory in another project.

Reply:
Going on the subject, I'm not sure I'm getting what to change here